nixos/doc/manual/development/writing-nixos-tests.section.md

```diff
@@ -417,3 +478,183 @@ import ./make-test-python.nix
+Laziness makes this pattern efficient.
+
+The test framework ([`runTest`](#sec-calling-nixos-tests)) takes care of
+traversing the choices one by one, constructing a tree of attribute sets where
+each edge represents a choice.
+
+A parameterized test may look as follows:
+
+```nix
+{ lib, backend, ... }: {
+  name = "foo-${backend}";
+
+  matrix.backend.choice.smtp.module = {};
+  matrix.backend.choice.rest.module = {};
+
+  defaults.services.foo.backend = backend;
+
+  nodes = …;
+  testScript = ''
+    ${lib.optionalString (backend == "rest") ''
+      # extra setup for the rest backend
+    ''}
+    # …
+  '';
+}
+```
+
+`runTest` transforms this to:
+
+```nix
+{
+  backend-rest = «derivation /nix/store/…-foo-rest.drv»;
+  backend-smtp = «derivation /nix/store/…-foo-smtp.drv»;
+
+  # make `nix-build` traverse this attrset to evaluate nested tests
+  recurseForDerivations = true;
+  # …
+}
+```
+
+It has picked up on the `matrix` and only returns the derivations constructed for each choice.
+
+`runTest` works by invoking the module system to lazily produce all option values, and then extracts the relevant values from it.
+This means that some options exist, but are never evaluated. For example, when `matrix` isn't empty, `nodes.<name>` is ignored in favor of `matrix.<decision>.choice.<choice>.module.nodes.<name>`.
+
+Nonetheless, the definitions that would contribute to `nodes.<name>` are useful, as they contribute to `matrix.<decision>.choice.<choice>.module.nodes.<name>`. The `module` option always duplicates and extends the root of the test configuration.
+
+Some variations aren't simply parametric, but require specific configuration for some choices.
+These can be added in the `module` submodule.
+
+The following example shows how tests in a test matrix can be quite different, while sharing common configuration.
+
+```nix
+{ lib, setup, ... }: {
+  matrix.backend.choice.smtp.module = { lib, ... }: {
+    defaults.services.foo.backend = "smtp";
+    nodes.smtpserver = { … };
+    _module.args.setup = "";
+  };
+  matrix.backend.choice.rest.module = { lib, ... }: {
+    _module.args.setup = ''
+      # extra setup for the rest backend
+      # …
+    '';
+  };
+
+  nodes = …;
+  testScript = ''
+    ${setup}
+    # …
+  '';
+}
+```
+
+The example shows the solutions to the following problems:
+ - The test cases require different configuration. For example, the `smtp` test sets the backend to `"smtp"` for all nodes, instead of relying on the default value.
+ - One test case requires an extra VM. `nodes.smtpserver` is only defined for the `smtp` test.
+ - One test case needs extra setup to be done in the `testScript`. It uses an ad hoc module argument called `setup` to achieve this.
+
+You may notice that the test appears to pull a `setup` value out of thin air, and indeed this would fail if `runTest` had accessed `testScript` directly instead of looking inside the `matrix.<...>.module` options. Thanks to laziness, this is not a problem.
```

Comment:
This does not explain at all why it works.

```suggestion
You may notice that the test appears to pull a `setup` value out of thin air.
Indeed, this would fail if `runTest` had accessed `testScript` directly instead of looking inside the `matrix.<...>.module` options.
Instead, `runTest` calls the top-level test module declaration in each of the configurations it produces, passing the `setup` argument with the value set in `_module.args.setup`.
```

Is this correct?

Reply:
A correct understanding, but with some incorrect technicalities.
 - "declaration" should be "definition"
 - we're giving `runTest` a bit too much agency, but if we need to mention all indirections, we're not more helpful than the code itself. This applies also to my own writing as well.

I'm changing it to this

```suggestion
You may notice that the test appears to pull a `setup` value out of thin air.
Indeed this would fail if `runTest` had accessed `testScript` directly instead of looking inside the `matrix.<...>.module` options.
In those `matrix.<...>.module` contexts, the definition of `testScript` is valid.
Thanks to laziness, the invalid `testScript` is never accessed and never becomes a problem.
```